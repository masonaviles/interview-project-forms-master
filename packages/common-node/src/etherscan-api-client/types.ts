import { Interface as ContractInterface } from 'ethers'

export interface IContractABIFetcher {
  getContractABI(address: string): Promise<ContractInterface>
}
