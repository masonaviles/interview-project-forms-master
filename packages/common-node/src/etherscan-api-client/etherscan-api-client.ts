import qs from 'querystring'
import axios, { AxiosInstance } from 'axios'
import { NetworkID } from 'common'
import { Interface as ContractInterface } from 'ethers'
import { IContractABIFetcher } from './types'

export class EtherscanAPIClient implements IContractABIFetcher {
  axios: AxiosInstance

  constructor(private readonly apiKey: string, private readonly network: NetworkID) {
    this.axios = axios.create({
      timeout: 10_000,
    })
    this.getAPIURL()
  }

  private getAPIURL() {
    switch (this.network) {
      case 'ethereum-mainnet':
        return 'https://api.etherscan.io/api'
      case 'optimism-mainnet':
        return 'https://api-optimistic.etherscan.io/api'
      case 'base-mainnet':
        return 'https://api.basescan.org/api'
      case 'base-goerli':
        return 'https://api-goerli.basescan.org/api'
      default:
        throw new Error(`Network ${this.network} unsupported by etherscan api for now`)
    }
  }

  private getRawContractABI(address: string) {
    const queryObject = {
      module: 'contract',
      action: 'getabi',
      address,
      apiKey: this.apiKey,
    }

    const apiURL = this.getAPIURL()

    return this.axios.get(`${apiURL}?${qs.stringify(queryObject)}`).then((resp) => {
      const resultAsString = resp.data.result
      return JSON.parse(resultAsString)
    })
  }

  public getContractABI(address: string) {
    return this.getRawContractABI(address).then(
      (result) => new ContractInterface(result)
    )
  }
}
