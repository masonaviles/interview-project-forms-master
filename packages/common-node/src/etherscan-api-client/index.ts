export * from './etherscan-api-client'
