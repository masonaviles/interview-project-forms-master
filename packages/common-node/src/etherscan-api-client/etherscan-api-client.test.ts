import { beforeAll, describe, expect, it } from 'vitest'
import { EtherscanAPIClient } from './etherscan-api-client'

describe.skip(EtherscanAPIClient, async () => {
  let etherscanAPIClient: EtherscanAPIClient

  beforeAll(() => {
    if (!process.env.ETHERSCAN_API_KEY) {
      throw new Error('Expected process.env.ETHERSCAN_API_KEY to be defined')
    }
    etherscanAPIClient = new EtherscanAPIClient(
      process.env.ETHERSCAN_API_KEY,
      'ethereum-mainnet'
    )
  })

  describe('getContractABI', async () => {
    it('should behavior', async () => {
      const contractABI = await etherscanAPIClient.getContractABI(
        // bitart
        '0xd70f41dd5875eee7fa9dd8048567bc932124a8d2'
      )
      expect(contractABI).toMatchSnapshot()
    })
  })
})
