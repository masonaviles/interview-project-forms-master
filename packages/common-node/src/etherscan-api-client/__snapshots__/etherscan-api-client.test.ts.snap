// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EtherscanAPIClient > getContractABI > should behavior 1`] = `
Interface {
  "deploy": ConstructorFragment {
    "gas": null,
    "inputs": [
      ParamType {
        "arrayChildren": null,
        "arrayLength": null,
        "baseType": "address",
        "components": null,
        "indexed": undefined,
        "name": "_feesTo",
        "type": "address",
      },
    ],
    "payable": false,
    "type": "constructor",
  },
  "fallback": null,
  "fragments": [
    FunctionFragment {
      "constant": true,
      "gas": null,
      "inputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "bytes4",
          "components": null,
          "indexed": undefined,
          "name": "interfaceId",
          "type": "bytes4",
        },
      ],
      "name": "supportsInterface",
      "outputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "bool",
          "components": null,
          "indexed": undefined,
          "name": "",
          "type": "bool",
        },
      ],
      "payable": false,
      "stateMutability": "view",
      "type": "function",
    },
    FunctionFragment {
      "constant": true,
      "gas": null,
      "inputs": [],
      "name": "name",
      "outputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "string",
          "components": null,
          "indexed": undefined,
          "name": "",
          "type": "string",
        },
      ],
      "payable": false,
      "stateMutability": "view",
      "type": "function",
    },
    FunctionFragment {
      "constant": true,
      "gas": null,
      "inputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "uint256",
          "components": null,
          "indexed": undefined,
          "name": "tokenId",
          "type": "uint256",
        },
      ],
      "name": "getApproved",
      "outputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": undefined,
          "name": "",
          "type": "address",
        },
      ],
      "payable": false,
      "stateMutability": "view",
      "type": "function",
    },
    FunctionFragment {
      "constant": false,
      "gas": null,
      "inputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": undefined,
          "name": "to",
          "type": "address",
        },
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "uint256",
          "components": null,
          "indexed": undefined,
          "name": "tokenId",
          "type": "uint256",
        },
      ],
      "name": "approve",
      "outputs": [],
      "payable": false,
      "stateMutability": "nonpayable",
      "type": "function",
    },
    FunctionFragment {
      "constant": true,
      "gas": null,
      "inputs": [],
      "name": "totalSupply",
      "outputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "uint256",
          "components": null,
          "indexed": undefined,
          "name": "",
          "type": "uint256",
        },
      ],
      "payable": false,
      "stateMutability": "view",
      "type": "function",
    },
    FunctionFragment {
      "constant": false,
      "gas": null,
      "inputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "uint256",
          "components": null,
          "indexed": undefined,
          "name": "tokenId",
          "type": "uint256",
        },
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "string",
          "components": null,
          "indexed": undefined,
          "name": "hash",
          "type": "string",
        },
      ],
      "name": "mintPhoto",
      "outputs": [],
      "payable": false,
      "stateMutability": "nonpayable",
      "type": "function",
    },
    FunctionFragment {
      "constant": false,
      "gas": null,
      "inputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": undefined,
          "name": "from",
          "type": "address",
        },
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": undefined,
          "name": "to",
          "type": "address",
        },
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "uint256",
          "components": null,
          "indexed": undefined,
          "name": "tokenId",
          "type": "uint256",
        },
      ],
      "name": "transferFrom",
      "outputs": [],
      "payable": false,
      "stateMutability": "nonpayable",
      "type": "function",
    },
    FunctionFragment {
      "constant": false,
      "gas": null,
      "inputs": [],
      "name": "withdrawFunds",
      "outputs": [],
      "payable": false,
      "stateMutability": "nonpayable",
      "type": "function",
    },
    FunctionFragment {
      "constant": false,
      "gas": null,
      "inputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "uint256",
          "components": null,
          "indexed": undefined,
          "name": "tokenId",
          "type": "uint256",
        },
      ],
      "name": "buyToken",
      "outputs": [],
      "payable": true,
      "stateMutability": "payable",
      "type": "function",
    },
    FunctionFragment {
      "constant": true,
      "gas": null,
      "inputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": undefined,
          "name": "owner",
          "type": "address",
        },
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "uint256",
          "components": null,
          "indexed": undefined,
          "name": "index",
          "type": "uint256",
        },
      ],
      "name": "tokenOfOwnerByIndex",
      "outputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "uint256",
          "components": null,
          "indexed": undefined,
          "name": "",
          "type": "uint256",
        },
      ],
      "payable": false,
      "stateMutability": "view",
      "type": "function",
    },
    FunctionFragment {
      "constant": false,
      "gas": null,
      "inputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": undefined,
          "name": "from",
          "type": "address",
        },
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": undefined,
          "name": "to",
          "type": "address",
        },
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "uint256",
          "components": null,
          "indexed": undefined,
          "name": "tokenId",
          "type": "uint256",
        },
      ],
      "name": "safeTransferFrom",
      "outputs": [],
      "payable": false,
      "stateMutability": "nonpayable",
      "type": "function",
    },
    FunctionFragment {
      "constant": true,
      "gas": null,
      "inputs": [],
      "name": "totalBought",
      "outputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "uint256",
          "components": null,
          "indexed": undefined,
          "name": "",
          "type": "uint256",
        },
      ],
      "payable": false,
      "stateMutability": "view",
      "type": "function",
    },
    FunctionFragment {
      "constant": true,
      "gas": null,
      "inputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "uint256",
          "components": null,
          "indexed": undefined,
          "name": "index",
          "type": "uint256",
        },
      ],
      "name": "tokenByIndex",
      "outputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "uint256",
          "components": null,
          "indexed": undefined,
          "name": "",
          "type": "uint256",
        },
      ],
      "payable": false,
      "stateMutability": "view",
      "type": "function",
    },
    FunctionFragment {
      "constant": true,
      "gas": null,
      "inputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "uint256",
          "components": null,
          "indexed": undefined,
          "name": "tokenId",
          "type": "uint256",
        },
      ],
      "name": "ownerOf",
      "outputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": undefined,
          "name": "",
          "type": "address",
        },
      ],
      "payable": false,
      "stateMutability": "view",
      "type": "function",
    },
    FunctionFragment {
      "constant": true,
      "gas": null,
      "inputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": undefined,
          "name": "owner",
          "type": "address",
        },
      ],
      "name": "balanceOf",
      "outputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "uint256",
          "components": null,
          "indexed": undefined,
          "name": "",
          "type": "uint256",
        },
      ],
      "payable": false,
      "stateMutability": "view",
      "type": "function",
    },
    FunctionFragment {
      "constant": false,
      "gas": null,
      "inputs": [],
      "name": "renounceOwnership",
      "outputs": [],
      "payable": false,
      "stateMutability": "nonpayable",
      "type": "function",
    },
    FunctionFragment {
      "constant": true,
      "gas": null,
      "inputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "uint256",
          "components": null,
          "indexed": undefined,
          "name": "",
          "type": "uint256",
        },
      ],
      "name": "tokenHashes",
      "outputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "string",
          "components": null,
          "indexed": undefined,
          "name": "",
          "type": "string",
        },
      ],
      "payable": false,
      "stateMutability": "view",
      "type": "function",
    },
    FunctionFragment {
      "constant": true,
      "gas": null,
      "inputs": [],
      "name": "owner",
      "outputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": undefined,
          "name": "",
          "type": "address",
        },
      ],
      "payable": false,
      "stateMutability": "view",
      "type": "function",
    },
    FunctionFragment {
      "constant": true,
      "gas": null,
      "inputs": [],
      "name": "isOwner",
      "outputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "bool",
          "components": null,
          "indexed": undefined,
          "name": "",
          "type": "bool",
        },
      ],
      "payable": false,
      "stateMutability": "view",
      "type": "function",
    },
    FunctionFragment {
      "constant": false,
      "gas": null,
      "inputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": undefined,
          "name": "to",
          "type": "address",
        },
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "bool",
          "components": null,
          "indexed": undefined,
          "name": "approved",
          "type": "bool",
        },
      ],
      "name": "setApprovalExternal",
      "outputs": [],
      "payable": false,
      "stateMutability": "nonpayable",
      "type": "function",
    },
    FunctionFragment {
      "constant": true,
      "gas": null,
      "inputs": [],
      "name": "symbol",
      "outputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "string",
          "components": null,
          "indexed": undefined,
          "name": "",
          "type": "string",
        },
      ],
      "payable": false,
      "stateMutability": "view",
      "type": "function",
    },
    FunctionFragment {
      "constant": true,
      "gas": null,
      "inputs": [],
      "name": "currentPrice",
      "outputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "uint256",
          "components": null,
          "indexed": undefined,
          "name": "",
          "type": "uint256",
        },
      ],
      "payable": false,
      "stateMutability": "view",
      "type": "function",
    },
    FunctionFragment {
      "constant": false,
      "gas": null,
      "inputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": undefined,
          "name": "to",
          "type": "address",
        },
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "bool",
          "components": null,
          "indexed": undefined,
          "name": "approved",
          "type": "bool",
        },
      ],
      "name": "setApprovalForAll",
      "outputs": [],
      "payable": false,
      "stateMutability": "nonpayable",
      "type": "function",
    },
    FunctionFragment {
      "constant": false,
      "gas": null,
      "inputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": undefined,
          "name": "from",
          "type": "address",
        },
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": undefined,
          "name": "to",
          "type": "address",
        },
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "uint256",
          "components": null,
          "indexed": undefined,
          "name": "tokenId",
          "type": "uint256",
        },
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "bytes",
          "components": null,
          "indexed": undefined,
          "name": "_data",
          "type": "bytes",
        },
      ],
      "name": "safeTransferFrom",
      "outputs": [],
      "payable": false,
      "stateMutability": "nonpayable",
      "type": "function",
    },
    FunctionFragment {
      "constant": true,
      "gas": null,
      "inputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "uint256",
          "components": null,
          "indexed": undefined,
          "name": "tokenId",
          "type": "uint256",
        },
      ],
      "name": "tokenURI",
      "outputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "string",
          "components": null,
          "indexed": undefined,
          "name": "",
          "type": "string",
        },
      ],
      "payable": false,
      "stateMutability": "view",
      "type": "function",
    },
    FunctionFragment {
      "constant": true,
      "gas": null,
      "inputs": [],
      "name": "baseTokenURI",
      "outputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "string",
          "components": null,
          "indexed": undefined,
          "name": "",
          "type": "string",
        },
      ],
      "payable": false,
      "stateMutability": "pure",
      "type": "function",
    },
    FunctionFragment {
      "constant": true,
      "gas": null,
      "inputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": undefined,
          "name": "owner",
          "type": "address",
        },
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": undefined,
          "name": "operator",
          "type": "address",
        },
      ],
      "name": "isApprovedForAll",
      "outputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "bool",
          "components": null,
          "indexed": undefined,
          "name": "",
          "type": "bool",
        },
      ],
      "payable": false,
      "stateMutability": "view",
      "type": "function",
    },
    FunctionFragment {
      "constant": true,
      "gas": null,
      "inputs": [],
      "name": "getCurrentPrice",
      "outputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "uint256",
          "components": null,
          "indexed": undefined,
          "name": "",
          "type": "uint256",
        },
      ],
      "payable": false,
      "stateMutability": "view",
      "type": "function",
    },
    FunctionFragment {
      "constant": false,
      "gas": null,
      "inputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": undefined,
          "name": "newOwner",
          "type": "address",
        },
      ],
      "name": "transferOwnership",
      "outputs": [],
      "payable": false,
      "stateMutability": "nonpayable",
      "type": "function",
    },
    ConstructorFragment {
      "gas": null,
      "inputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": undefined,
          "name": "_feesTo",
          "type": "address",
        },
      ],
      "payable": false,
      "type": "constructor",
    },
    EventFragment {
      "anonymous": false,
      "inputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "uint256",
          "components": null,
          "indexed": true,
          "name": "tokenId",
          "type": "uint256",
        },
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": true,
          "name": "acquirer",
          "type": "address",
        },
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "uint256",
          "components": null,
          "indexed": false,
          "name": "price",
          "type": "uint256",
        },
      ],
      "name": "TokenSold",
      "type": "event",
    },
    EventFragment {
      "anonymous": false,
      "inputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": true,
          "name": "previousOwner",
          "type": "address",
        },
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": true,
          "name": "newOwner",
          "type": "address",
        },
      ],
      "name": "OwnershipTransferred",
      "type": "event",
    },
    EventFragment {
      "anonymous": false,
      "inputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": true,
          "name": "from",
          "type": "address",
        },
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": true,
          "name": "to",
          "type": "address",
        },
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "uint256",
          "components": null,
          "indexed": true,
          "name": "tokenId",
          "type": "uint256",
        },
      ],
      "name": "Transfer",
      "type": "event",
    },
    EventFragment {
      "anonymous": false,
      "inputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": true,
          "name": "owner",
          "type": "address",
        },
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": true,
          "name": "approved",
          "type": "address",
        },
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "uint256",
          "components": null,
          "indexed": true,
          "name": "tokenId",
          "type": "uint256",
        },
      ],
      "name": "Approval",
      "type": "event",
    },
    EventFragment {
      "anonymous": false,
      "inputs": [
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": true,
          "name": "owner",
          "type": "address",
        },
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "address",
          "components": null,
          "indexed": true,
          "name": "operator",
          "type": "address",
        },
        ParamType {
          "arrayChildren": null,
          "arrayLength": null,
          "baseType": "bool",
          "components": null,
          "indexed": false,
          "name": "approved",
          "type": "bool",
        },
      ],
      "name": "ApprovalForAll",
      "type": "event",
    },
  ],
  "receive": false,
}
`;
